presto-main/src/main/java/io/prestosql/operator/unnest/ReplicatedBlockBuilder.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator.unnest;
+
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.block.DictionaryBlock;
+
+import java.util.Arrays;
+
+import static com.google.common.base.Preconditions.checkElementIndex;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class manages the details for building replicate channel blocks without copying input data
+ */
+class ReplicatedBlockBuilder
+{
+    private Block source;
+    private int[] ids;
+    private int positionCount;
+
+    public ReplicatedBlockBuilder() {}
+
+    public void resetInputBlock(Block block)
+    {
+        this.source = requireNonNull(block, "block is null");
+    }
+
+    public void startNewOutput(int expectedEntries)
+    {
+        checkState(source != null, "source is null");
+        this.ids = new int[expectedEntries];
+        this.positionCount = 0;
+    }
+
+    /**
+     * Repeat the source element at position {@code index} for {@code count} times in the output
+     */
+    public void appendRepeated(int index, int count)
+    {
+        checkState(source != null, "source is null");
+        checkElementIndex(index, source.getPositionCount());
+
+        for (int i = 0; i < count; i++) {
```

Comment:
I implemented this:

```
        if (positionCount + count >= ids.length) {
            // Grow capacity
            int newSize = Math.max(calculateNewArraySize(ids.length), positionCount + count);
            ids = Arrays.copyOf(ids, newSize);
        }
```

There is an edge case if `positionCount + count` is greater than UnnestOperatorBlockUtil.MAX_ARRAY_SIZE (or overflows). I don't expect it to be common though, since it would require a trillion elements in the block, hence ~billions in a row (and will probably fail before it comes to this). Do you think we still need to sanity-check it?

Reply:
In this case it would overflow to negative, and you would get only the result of `calculateNewArraySize(ids.length)`.  If that was not big enough, you would get an error when calling array fill.  Anyway, I don' think this is worth changing here.  In the future, you can use `addExact` instead of `+` to cause an exception to be thrown on overflow.